Merge type stubs for torch.nn.parallel

ringohoffman opened the pull request:
Fixes #91648

As explained in the tracking issue, the incomplete type stubs in `torch/nn/parallel` mask `DataParallel` methods relevant for subclassing and also mask type issues present in the code as well.

There are various renamings and `type: ignore` comments added to bypass errors that arose from the merging.

@ezyang

Commit:
Merge type stubs for torch.nn.parallel

Resolves #91648

Commit:
Fix torch.jit circular dependency

Skylion007 commented on torch/nn/parallel/scatter_gather.py:
```diff
@@ -43,11 +59,17 @@ def scatter_map(obj):
 
 
-def scatter_kwargs(inputs, kwargs, target_gpus, dim=0):
+# TODO More precise types here.
```
FYI: https://github.com/Instagram/MonkeyType can help you find more precise types here (if you remove the Any annotations).

Commit:
Define public interface for parallel_apply and replicate

ezyang commented:
added albanD for public API errors

ezyang reviewed (approved):
Seems fine, but you'll have to figure out how to shut up the public API errors. Probably a missing `__all__` somewhere.

Commit:
Improve scatter_kwargs typing

Commit:
Remove unnecessary import alias

The redundant import import alias defines a public interface, but this is already accomplished by inclusion in __all__, which all the imported variables are in

Commit:
Make get_a_var private to follow public interface rules

ringohoffman commented:
Looks like the BC linter didn't like changing `get_a_var` to be private to appease the public interface linter (https://github.com/pytorch/pytorch/pull/101528/commits/ceeb242bc7a8291c2f108659e46fd81d418380ec). LMK if we can suppress this or if you want to include `get_a_var` in `__all__` in `torch/nn/parallel/parallel_apply.py` (it is only referenced in that file).

ringohoffman commented:
@ezyang anything outstanding here?

ringohoffman commented:
> It's still failing lintrunner, gotta fix the rest of the type errors in other files. You can do it locally with `lintrunner --all-files` since the reduced file heuristic doesn't work quite right for mypy.

@ezyang I addressed these errors here: https://github.com/pytorch/pytorch/pull/102046

Commit:
Merge branch 'main' into merge-type-stubs-torch-nn-parallel

ezyang commented:
@pytorchbot merge

kit1980 commented:
@pytorchbot revert -m "Broke inductor tests https://github.com/pytorch/pytorch/actions/runs/5071348299/jobs/9107880424 ImportError: cannot import name 'get_a_var' from 'torch.nn.parallel.parallel_apply'" -c ignoredsignal

kit1980 commented:
The bc_linter actually flagged this on the PR:

GitHub Actions
/ bc_linter
Function get_a_var: function deleted

ringohoffman commented:
@kit1980 #102194
